kotlin/src/main/kotlin/com/linecorp/armeria/internal/common/kotlin/FlowCollectingPublisher.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License
+ */
+
+package com.linecorp.armeria.internal.common.kotlin
+
+import com.linecorp.armeria.common.stream.DefaultStreamMessage
+import kotlinx.coroutines.AbstractCoroutine
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.InternalCoroutinesApi
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.future.await
+import kotlinx.coroutines.handleCoroutineException
+import org.reactivestreams.Publisher
+import org.reactivestreams.Subscriber
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.EmptyCoroutineContext
+
+/**
+ * [Publisher] implementation which emits values collected from [Flow].
+ * Reactive streams back-pressure works on its backing [flow] too.
+ */
+@OptIn(InternalCoroutinesApi::class)
+internal class FlowCollectingPublisher<T : Any>(
+    private val flow: Flow<T>,
+    context: CoroutineContext = EmptyCoroutineContext
+) : Publisher<T>, AbstractCoroutine<Unit>(context, initParentJob = false, active = true) {
+    private val delegate = DefaultStreamMessage<T>()
+
+    override fun subscribe(s: Subscriber<in T>) {
+        this.start(CoroutineStart.DEFAULT, this) {
+            flow.collect {
+                delegate.write(it)
+                delegate.whenConsumed().await()
+            }
+            delegate.close()
+        }
+        delegate.whenComplete().handle { _, _ -> cancel() }
```

Comment:
Should we check completion before calling `cancel()`?
```kt
if (!isCompleted) {
   cancel()
}
```

Reply:
I think we are good to go as it is?

Official `kotlinx-coroutines` library just calls `cancel()` in a similar situation.
I think it's because coroutine has many states and cancellation should be handled differently depending on the state.

Ref.
https://github.com/Kotlin/kotlinx.coroutines/blob/master/integration/kotlinx-coroutines-jdk8/src/future/Future.kt#L43
https://github.com/Kotlin/kotlinx.coroutines/blob/master/integration/kotlinx-coroutines-jdk8/src/future/Future.kt#L43
https://github.com/Kotlin/kotlinx.coroutines/blob/65e1b84f742a4fff14c67a11a873a4977a0ba539/kotlinx-coroutines-core/common/src/JobSupport.kt#L661-L682

Reply:
If so, how about explicitly passing a singleton `CancellationException` so that we don't create an additional object?

Reply:
Otherwise, could we use `ctx.whenRequestCancelling()`?

Reply:
Oh, I think you're write about this (like you always are 😄).
Cancellation is a long process and we should avoid calling it if possible.

What about checking `isActive` flag?
```
 * A job has the following states:
 *
 * | **State**                        | [isActive] | [isCompleted] | [isCancelled] |
 * | -------------------------------- | ---------- | ------------- | ------------- |
 * | _New_ (optional initial state)   | `false`    | `false`       | `false`       |
 * | _Active_ (default initial state) | `true`     | `false`       | `false`       |
 * | _Completing_ (transient state)   | `true`     | `false`       | `false`       |
 * | _Cancelling_ (transient state)   | `false`    | `false`       | `true`        |
 * | _Cancelled_ (final state)        | `false`    | `true`        | `true`        |
 * | _Completed_ (final state)        | `false`    | `true`        | `false`       |
 *
...
 * ```
 *                                       wait children
 * +-----+ start  +--------+ complete   +-------------+  finish  +-----------+
 * | New | -----> | Active | ---------> | Completing  | -------> | Completed |
 * +-----+        +--------+            +-------------+          +-----------+
 *                  |  cancel / fail       |
 *                  |     +----------------+
 *                  |     |
 *                  V     V
 *              +------------+                           finish  +-----------+
 *              | Cancelling | --------------------------------> | Cancelled |
 *              +------------+                                   +-----------+
 * ```
```

Reply:
`isActive` also sounds good to me.